### ISSUE TITLE
Implement BalancedRandomForestClassifier

### ISSUE THREAD

**Issue by damienlancry** [haifengl/smile#600]
In Python, I am used to using the imbalanced-learn library for imbalanced datasets. That library simply builds upon scikit-learn. I am particularly interested in the BalancedRandomForestClassifier. 

This is a random forest that only resamples the majority class, so that all decision trees of the forest are trained on balanced bootstrapped samples of the original dataset.

Is there any plan to implement this algorithm in smile? otherwise could it be added to the roadmap? I would potentially be interested in contributing if you think that would be relevant to add this algorithm to smile.

More generally, the BalancedBaggingClassifier could also be implemented.

References:
https://imbalanced-learn.org/stable/generated/imblearn.ensemble.BalancedBaggingClassifier.html
https://imbalanced-learn.org/stable/generated/imblearn.ensemble.BalancedRandomForestClassifier.html
https://statistics.berkeley.edu/sites/default/files/tech-reports/666.pdf

**Comment by haifengl** [comment]
We use the parameter `classWeight` to balance data. Says you have two class in 10:1 ratio, set `classWeight` to `[10, 1]` or try other values.

**Comment by damienlancry** [comment]
Thanks @haifengl for the quick answer! Amazing indeed I can use this parameter!

In my use case, the minority class has 5 samples and the majority class has 9175 samples in the train set, the minority class has 1000 samples and the majority class has 8180 samples in the test set.

I use the balanced random forest with 100 trees, without bootstrapping (sampling without replacement) because I want each tree in the forest to be trained on all positive examples.

I compared the performance of imblearn to the performance of smile in a spylon notebook.

The code looks like this:
In:
```
%%init_spark
launcher.jars = ["/temporary/smile-scala_2.12-2.5.3.jar"]
launcher.packages = ["com.github.haifengl:smile-scala_2.12:2.5.3"]

import scala.language.postfixOps
import org.apache.spark.mllib.tree.model.RandomForestModel
import smile._
import smile.classification.randomForest
import smile.data._
import smile.data.formula._
import smile.data.measure._
import smile.validation._

def mean(list: List[Double]): Double = list.sum / list.size
def std(xs: List[Double]): Double = {
    val avg: Double = mean(xs)
    scala.math.sqrt(xs.map(xi =>scala.math.pow(xi - avg, 2)).sum / (xs.size))
}

val df_train = read.arff("data_train.arff")
val df_test = read.arff("data_test.arff")
val formula: Formula = "class" ~

%%time
val ytrue = df_test.select(600).toArray.map(_(0).toInt)
val gmeansAndFscore = ( 1 to 10 ).toList.map { int =>
    val rf = randomForest(formula, df_train, classWeight = Array(1835, 1))
    val ypred = Validation.test(rf, df_test)
    val tpr = sensitivity(ytrue, ypred)
    val tnr = specificity(ytrue, ypred)
    val ppv = precision(ytrue, ypred)
    val gmean = scala.math.sqrt(tpr * tnr)
    val fscore = 2 * ppv * tpr / (ppv + tpr)
    (tpr, tnr, ppv, gmean, fscore)
}
val tprs = gmeansAndFscore.map{ case (tpr, tnr, ppv, gmean, fscore) => tpr }
val tnrs = gmeansAndFscore.map{ case (tpr, tnr, ppv, gmean, fscore) => tnr }
val ppvs = gmeansAndFscore.map{ case (tpr, tnr, ppv, gmean, fscore) => ppv }
val gmeans = gmeansAndFscore.map{ case (tpr, tnr, ppv, gmean, fscore) => gmean }
val fscores = gmeansAndFscore.map{ case (tpr, tnr, ppv, gmean, fscore) => fscore }
println("tpr    = " + mean(tprs).toString + " ± " + std(tprs).toString)
println("tnr    = " + mean(tnrs).toString + " ± " + std(tnrs).toString)
println("ppv    = " + mean(ppvs).toString + " ± " + std(ppvs).toString)
println("gmean  = " + mean(gmeans).toString + " ± " + std(gmeans).toString)
println("fscore = " + mean(fscores).toString + " ± " + std(fscores).toString)
```
Out:
```
tpr    = 0.9135270541082166 ± 0.018884785249380143
tnr    = 0.9937790271327304 ± 0.0027095543564178244
ppv    = 0.94781458750958 ± 0.020883086365091067
gmean  = 0.9527486569633323 ± 0.00927138452800075
fscore = 0.9300306285301512 ± 0.009986452610368282
Time: 13.588037490844727 seconds.
```
And the python equivalent:
In:
```
import numpy as np
from numpy import mean, std
import pandas as pd
from scipy.io.arff import loadarff
from imblearn.ensemble import BalancedRandomForestClassifier
from imblearn.metrics import sensitivity_score, specificity_score
from sklearn.metrics import precision_score

def read_arff(path):
    raw_data = loadarff(path)
    df_data = pd.DataFrame(raw_data[0])
    df_data["class"] = np.where(df_data["class"] == b'1', 1, 0)
    return df_data

df_train = read_arff("data_train.arff")
df_test = read_arff("data_test.arff")

%%time
tprs = []
tnrs = []
ppvs = []
gmeans = []
fscores = []
for i in range(100):
    rf = BalancedRandomForestClassifier(bootstrap=False)
    rf.fit(df_train[df_train.columns.difference(["class"])], df_train["class"])
    ypred = rf.predict(df_test[df_test.columns.difference(["class"])])
    ytrue = df_test["class"]
    tpr = sensitivity_score(ytrue, ypred)
    tnr = specificity_score(ytrue, ypred)
    ppv = precision_score(ytrue, ypred)
    gmean = np.sqrt(tpr * tnr)
    fscore = 2 * ppv * tpr / (ppv + tpr)
    tprs.append(tpr)
    tnrs.append(tnr)
    ppvs.append(ppv)
    gmeans.append(gmean)
    fscores.append(fscore)
print("tpr    = " + str(mean(tprs)) + " ± " + str(std(tprs)))
print("tnr    = " + str(mean(tnrs)) + " ± " + str(std(tnrs)))
print("ppv    = " + str(mean(ppvs)) + " ± " + str(std(ppvs)))
print("gmean  = " + str(mean(gmeans)) + " ± " + str(std(gmeans)))
print("fscore = " + str(mean(fscores)) + " ± " + str(std(fscores)))
```
Out:
```
tpr    = 0.9629058116232463 ± 0.011259047809012636
tnr    = 0.9772622830603765 ± 0.010049370901310473
ppv    = 0.842156774604022 ± 0.05746163444094017
gmean  = 0.9700150891120947 ± 0.0056735396032671
fscore = 0.8972492214353851 ± 0.0314637069718618
CPU times: user 58.4 s, sys: 1.58 s, total: 60 s
Wall time: 59.9 s
```

The metric I care the most about is the geometric mean of the sensitivity and specificity. as you can see it is a bit lower for smile in this example. I am interested in understanding why smile and imblearn perform differently in this example. The only difference I am aware of in this example is the fact that imblearn samples without replacement whereas smile samples with replacement. I reran the above piece of python code by turning on the "bootstrap" parameter of the imblearn BalancedRandomForest and the output became: 
```
tpr    = 0.9640180360721443 ± 0.016085266808167235
tnr    = 0.9648337814715228 ± 0.01915372579607278
ppv    = 0.7819158121080276 ± 0.09114102020640531
gmean  = 0.9642883428876935 ± 0.0070912601593385695
fscore = 0.8598686955659276 ± 0.05213847170440634
CPU times: user 1min 20s, sys: 2.88 s, total: 1min 23s
Wall time: 1min 25s
```
As expected the geometric mean decreases a bit , but still not as low as smile. Is there another difference between smile implementation and sklearn implementation that I am unaware of?

I took a closer look at the code for RandomForest fit method and I have a question about this part:
```
            if (subsample == 1.0) {
                // Training samples draw with replacement.
                for (int i = 0; i < k; i++) {
                    // We used to do up sampling.
                    // But we switch to down sampling, which seems producing better AUC.
                    int ni = count[i];
                    int size = ni / weight[i];
                    int[] yj = yi[i];
                    for (int j = 0; j < size; j++) {
                        int xj = MathEx.randomInt(ni);
                        samples[yj[xj]] += 1; //classWeight[i];
                    }
                }
            } else {
                // Training samples draw without replacement.
                for (int i = 0; i < k; i++) {
                    // We used to do up sampling.
                    // But we switch to down sampling, which seems producing better AUC.
                    int size = (int) Math.round(subsample * count[i] / weight[i]);
                    int[] yj = yi[i];
                    int[] permutation = MathEx.permutate(count[i]);
                    for (int j = 0; j < size; j++) {
                        int xj = permutation[j];
                        samples[yj[xj]] += 1; //classWeight[i];
                    }
                }
            }
```
In my case I would be interested in sampling without replacement despite having a "subsample" rate of 1.0
Do you think it would be a good idea to do a pr to authorise this behaviour? it does not look like too much work and I could contribute if you think it would be relevant.

EDIT:
I actually managed to have the behaviour I wanted by setting subsample to 0.9999 (even if it is a bit of a hack).
the results are then :
```
tpr    = 0.9658617234468935 ± 0.010849677004521632
tnr    = 0.9761464189684674 ± 0.00950440451895059
ppv    = 0.8354015419172747 ± 0.05332109664228462
gmean  = 0.9709512861493237 ± 0.005343652812476201
fscore = 0.8948149667662616 ± 0.02929484868763466
Time: 141.68998456001282 seconds.
```
So exactly the same as imblearn which is very nice.

**Comment by damienlancry** [comment]
Amazing that works perfectly.

on a completely different matter, I read in the readme that it should be possible to reuse the model directly in spark, but I could not find any example or further documentation about that. 
I naively tried :
```
import org.apache.spark.mllib.tree.model.RandomForestModel
import smile._
import smile.classification.randomForest

val rf = randomForest(formula, df, classWeight = Array(1835, 1))
write(rf, "rf")
val randomForest = RandomForestModel.load(sc, "rf")
```
but it obviously failed.

Could you please give some further directions about how to load models and reuse them in spark?

**Comment by haifengl** [comment]
smile and mllib are different things. To use smile in spark, just load it as a smile object and use it as a lambda.

**Comment by damienlancry** [comment]
ok thanks!
about the balanced random forest: In my case I deal with extremely imbalanced datasets for binary classification (5 positive examples, 10 000 negative examples). 
Therefore It is important in this case that all trees are trained on all positive examples (sampling without replacement, no subsampling). 
However in smile, if you want to do sampling without replacement, you have to do subsampling (subsample != 1.0).
I managed to have the behaviour I wanted (no subsampling, but sampling without replacement) by setting subsample to 0.9999.
However this feels a bit hacky. maybe it would be better to change the code of RandomForest fit method to decouple bootstrapping (sampling with/without replacement) and subsampling.

I played with the code and I added a parameter boolean bootstrap to the fit method and I think this would work:
```
            // Stratified sampling in case that class is unbalanced.
            // That is, we sample each class separately.
            if (bootstrap) {
                // Training samples draw with replacement.
                for (int i = 0; i < k; i++) {
                    // We used to do up sampling.
                    // But we switch to down sampling, which seems producing better AUC.
                    int ni = count[i];
                    int size = (int) Math.round(subsample * ni / weight[i]);
                    int[] yj = yi[i];
                    for (int j = 0; j < size; j++) {
                        int xj = MathEx.randomInt(ni);
                        samples[yj[xj]] += 1; //classWeight[i];
                    }
                }
            } else {
                // Training samples draw without replacement.
                for (int i = 0; i < k; i++) {
                    // We used to do up sampling.
                    // But we switch to down sampling, which seems producing better AUC.
                    int size = (int) Math.round(subsample * count[i] / weight[i]);
                    int[] yj = yi[i];
                    int[] permutation = MathEx.permutate(count[i]);
                    for (int j = 0; j < size; j++) {
                        int xj = permutation[j];
                        samples[yj[xj]] += 1; //classWeight[i];
                    }
                }
            }
```
the code is exactly the same as it is right now, except the if condition is on the bootstrap boolean, and if it is true the sampling is done with replacement with a size of (int) Math.round(subsample * ni / weight[i]);
The else condition is the exact same as before.

I could open a pr with these changes if you think it could bring value to smile.

**Comment by haifengl** [comment]
Thanks. Random forest already has many hyper-parameters. I am reluctant to add one more for a corner case that has a workaround.

**Comment by damienlancry** [comment]
all right then, thank you for your help!
